In [1]:
from utilities.utils_clean import *
from utilities.test_dataset_utils import *

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\amoha\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\amoha\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:

#download_patents_pto(start_year=2015, end_year=2015, kind='grant',download_path="data")
#unzip_files("data","patent_grants_2015")

In [2]:
# Example usage
test_dataset = create_test_dataset_from_freilich( #"patent_xmls_2010.json"
    year=2015,  # The year you want to analyze
    freilich_data_path="Freilich.Data.Compressed.xlsb",  # Path to your Freilich dataset
    path_to_all_xmls_for_chosen_year="../app/data/patent_grants_2015"  # Directory containing XML files
)

# # Optional: Check the results
# print(f"Number of patents extracted: {len(test_dataset)}")
# print("Sample document numbers:", list(test_dataset.keys())[:5])

Processing ipg150106.xml
Processing ipg150113.xml
Processing ipg150120.xml
Processing ipg150127.xml
Processing ipg150203.xml
Processing ipg150210.xml
Processing ipg150217.xml
Processing ipg150224.xml
Processing ipg150303.xml
Processing ipg150310.xml
Processing ipg150317.xml
Processing ipg150324.xml
Processing ipg150331.xml
Processing ipg150407.xml
Processing ipg150414.xml
Processing ipg150421.xml
Processing ipg150428.xml
Processing ipg150505.xml
Processing ipg150512.xml
Processing ipg150519.xml
Processing ipg150526.xml
Processing ipg150602.xml
Processing ipg150609.xml
Processing ipg150616.xml
Processing ipg150623.xml
Processing ipg150630.xml
Processing ipg150707.xml
Processing ipg150714.xml
Processing ipg150721.xml
Processing ipg150728.xml
Processing ipg150804.xml
Processing ipg150811.xml
Processing ipg150818.xml
Processing ipg150825.xml
Processing ipg150901.xml
Processing ipg150908.xml
Processing ipg150915.xml
Processing ipg150922.xml
Processing ipg150929.xml
Processing ipg151006.xml


In [3]:
save_as_json(test_dataset, "patent_xmls_2015.json")

Saved as patent_xmls_2015.json


In [70]:
data = read_json("patent_xmls_2015.json")
df = pd.DataFrame.from_dict(data,orient="index").reset_index()
df.columns = ["patentnumber","xml"]
df.head()

patentnumber                                                xml
0     RE045323  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
1     RE045324  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
2     RE045325  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
3      8925349  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
4      8925551  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...

In [3]:
# test_data = load_from_pickle("../data/test_dataset_2015.pkl")

# df_test_data = pd.DataFrame(test_data,index=["xml"],).T.reset_index()
# df_test_data.columns = ["patentnumber","xml"] 


# df = read_xlsb_file()

# df[df.issueyear == 2015].to_csv("freilichdataet_2015.csv")
df2 = pd.read_csv("freilichdataet_2015.csv")
df2["patentnumber"] = df2["patentnumber"].astype(str).transform(lambda x: x.replace(".0", ""))
df2["patentnumber"] = df2["patentnumber"].apply(remove_leadiong_zeros)
df2.head()

Unnamed: 0 patentnumber  appyear  issueyear inventorcountry  divisional  \
0      497653      8969641   1976.0     2015.0              US         0.0   
1      497654      9134302   1995.0     2015.0              US         1.0   
2      497655      8926966   1995.0     2015.0              US         1.0   
3      497656      9181140   1993.0     2015.0              US         0.0   
4      497657      9199886   2009.0     2015.0              US         0.0   

   continuation  prioritydate  priority.formatted  priorityyear  ...  yr4  \
0           0.0    19760903.0             28006.0        1976.0  ...  NaN   
1           1.0    19830505.0             30441.0        1983.0  ...  NaN   
2           1.0    19900921.0             33137.0        1990.0  ...  NaN   
3           1.0    19920916.0             33863.0        1992.0  ...  0.0   
4           1.0    19940119.0             34353.0        1994.0  ...  0.0   

   yr8  yr11  claims wordaverage  litigated  orangebook  forward.cites  \
0  NaN   NaN     NaN         NaN        0.0         0.0            0.0   
1  NaN   NaN     NaN         NaN        0.0         0.0            0.0   
2  NaN   NaN     NaN         NaN        0.0         0.0            0.0   
3  0.0   0.0     NaN         NaN        0.0         0.0            0.0   
4  0.0   0.0     NaN         NaN        0.0         0.0            0.0   

   industry  subindustry  
0       1.0         19.0  
1       NaN          NaN  
2       3.0         31.0  
3       NaN          NaN  
4       NaN          NaN  

[5 rows x 32 columns]

In [25]:
import sqlite3 as sql
def test_algorithm(year):
    conn = sql.connect("../app/db/patents.db")
    cursor = conn.cursor()

    # Execute query and fetch results
    stats = cursor.execute("select * from patent_statistics").fetchall()
    stats_df = pd.DataFrame(stats, columns=["year", "patentnumber", "prophetic_patents", "nonprophetic_patents","unknown_patents"])
    conn.close()

    df_year = df[df.issueyear == year][["patentnumber","prophetic","nonprophetic","issueyear"]]
    df_year["patentnumber"] = df_year["patentnumber"].astype(str).transform(lambda x: x.replace(".0", ""))

    merged = stats_df.merge(df_year, on="patentnumber")
    print(len(merged))
    # final = dic_to_dic_w_tense_test(doc_w_exp,threshold=0)
    # df_final = pd.DataFrame(final).T.reset_index()
    df_final = merged
    #df_final.columns = ["patentnumber","past","present","Unknown"]
    df_final["patentnumber"] = df_final["patentnumber"].apply(remove_leadiong_zeros)
    df_check =merged
    #df_check["past"] = df_check["past"] + df_check["Unknown"]
    df_check["Total_Extracted"] = df_check["nonprophetic_patents"] + df_check["prophetic_patents"]  #+ df_check["Unknown"]
    df_check["Total_Freilich"] = df_check["prophetic"] + df_check["nonprophetic"]
    df_check["prophetic_error"] = np.sqrt((df_check["prophetic"] - df_check["prophetic_patents"])**2)
    df_check["nonprophetic_error"] = np.sqrt((df_check["nonprophetic"] - df_check["nonprophetic_patents"])**2)
    df_check["Total_Mean_error"] = np.sqrt((df_check["Total_Freilich"] - df_check["Total_Extracted"])**2)
    df_check["Sum_error"] = df_check["prophetic_error"] + df_check["nonprophetic_error"] + df_check["Total_Mean_error"]
    print(f"Number of exact matches: {len(df_check[(df_check.prophetic_error ==0) & (df_check.nonprophetic_error ==0) & (df_check.Total_Mean_error ==0)])} out of {len(df_check)} , Percentage: {len(df_check[(df_check.prophetic_error ==0) & (df_check.nonprophetic_error ==0) & (df_check.Total_Mean_error ==0)])/len(df_check)*100}")
    print(f"Number of exact num of patent extracted: {len(df_check[(df_check.Total_Extracted == df_check.Total_Freilich)])} out of {len(df_check)}")
    print(f"Avg Total error: {df_check['Total_Mean_error'].mean()}, num of corrects:  {df_check[df_check['Total_Mean_error'] == 0].shape[0]}")
    print(f"Avg Total prophetic error: {df_check['prophetic_error'].mean()}, num of corrects: {df_check[df_check['prophetic_error'] == 0].shape[0]}")
    print(f"Avg Total nonprophetic error: {df_check['nonprophetic_error'].mean()}, num of corrects: {df_check[df_check['nonprophetic_error'] == 0].shape[0]}")
    print(f"{df_check['Sum_error'].sum()}")
    print(f"Sum of number of Unknowns: {df_check['unknown_patents'].sum()}")
    print(f"Number of patets with experiments extracted {len(df_check)}")
    df_check.sort_values("Sum_error",ascending=False).head(20)
test_algorithm(2010)

327
Number of exact matches: 10 out of 327 , Percentage: 3.058103975535168
Number of exact num of patent extracted: 293 out of 327
Avg Total error: 0.7492354740061162, num of corrects:  293
Avg Total prophetic error: 17.122324159021407, num of corrects: 10
Avg Total nonprophetic error: 16.587155963302752, num of corrects: 11
11268.0
Sum of number of Unknowns: 0
Number of patets with experiments extracted 327


In [4]:
test_dataset = read_json("patent_xmls_2010.json")
test_dataset_df = pd.DataFrame.from_dict(test_dataset, orient='index').reset_index()
test_dataset_df.columns = ['patentnumber', 'xml']
test_dataset_df.head()

patentnumber                                                xml
0      7641702  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
1      7641704  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
2      7641709  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
3      7641721  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...
4      7641723  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...

In [6]:
merged = df.merge(df2, on="patentnumber")
print(len(merged))
merged.head()

25081


patentnumber                                                xml  Unnamed: 0  \
0     RE045323  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...      517229   
1     RE045324  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...      498769   
2     RE045325  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...      502690   
3      8925349  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...      522537   
4      8925551  <?xml version="1.0" encoding="UTF-8"?>\n<!DOCT...      521220   

   appyear  issueyear inventorcountry  divisional  continuation  prioritydate  \
0   2014.0     2015.0              DE         0.0           0.0    20090408.0   
1   2012.0     2015.0              US         0.0           0.0    20040603.0   
2   2013.0     2015.0              US         0.0           1.0    20080418.0   
3   2009.0     2015.0              CN         0.0           0.0    20090519.0   
4   2005.0     2015.0              US         0.0           0.0    20040226.0   

   priority.formatted  ...  yr4 yr8  yr11  claims  wordaverage litigated  \
0             39911.0  ...  NaN NaN   NaN     NaN          NaN       0.0   
1             38141.0  ...  NaN NaN   NaN     NaN          NaN       0.0   
2             39556.0  ...  NaN NaN   NaN     NaN          NaN       0.0   
3             39952.0  ...  NaN NaN   NaN     NaN          NaN       0.0   
4             38043.0  ...  NaN NaN   NaN     NaN          NaN       0.0   

   orangebook  forward.cites  industry  subindustry  
0         0.0            0.0       NaN          NaN  
1         0.0            0.0       NaN          NaN  
2         0.0            0.0       NaN          NaN  
3         0.0            0.0       5.0         51.0  
4         0.0            0.0       3.0         32.0  

[5 rows x 33 columns]

In [10]:
import numpy as np
# final = dic_to_dic_w_tense_test(doc_w_exp,threshold=0)
# df_final = pd.DataFrame(final).T.reset_index()
df_final = merged
#df_final.columns = ["patentnumber","past","present","Unknown"]
df_final["patentnumber"] = df_final["patentnumber"].apply(remove_leadiong_zeros)
df_check =merged
#df_check["past"] = df_check["past"] + df_check["Unknown"]
df_check["Total_Extracted"] = df_check["nonprophetic_patents"] + df_check["prophetic_patents"]  #+ df_check["Unknown"]
df_check["Total_Freilich"] = df_check["prophetic"] + df_check["nonprophetic"]
df_check["prophetic_error"] = np.sqrt((df_check["prophetic"] - df_check["prophetic_patents"])**2)
df_check["nonprophetic_error"] = np.sqrt((df_check["nonprophetic"] - df_check["nonprophetic_patents"])**2)
df_check["Total_Mean_error"] = np.sqrt((df_check["Total_Freilich"] - df_check["Total_Extracted"])**2)
df_check["Sum_error"] = df_check["prophetic_error"] + df_check["nonprophetic_error"] + df_check["Total_Mean_error"]
print(f"Number of exact matches: {len(df_check[(df_check.prophetic_error ==0) & (df_check.nonprophetic_error ==0) & (df_check.Total_Mean_error ==0)])} out of {len(df_check)} , Percentage: {len(df_check[(df_check.prophetic_error ==0) & (df_check.nonprophetic_error ==0) & (df_check.Total_Mean_error ==0)])/len(df_check)*100}")
print(f"Number of exact num of patent extracted: {len(df_check[(df_check.Total_Extracted == df_check.Total_Freilich)])} out of {len(df_check)}")
print(f"Avg Total error: {df_check['Total_Mean_error'].mean()}, num of corrects:  {df_check[df_check['Total_Mean_error'] == 0].shape[0]}")
print(f"Avg Total prophetic error: {df_check['prophetic_error'].mean()}, num of corrects: {df_check[df_check['prophetic_error'] == 0].shape[0]}")
print(f"Avg Total nonprophetic error: {df_check['nonprophetic_error'].mean()}, num of corrects: {df_check[df_check['nonprophetic_error'] == 0].shape[0]}")
print(f"{df_check['Sum_error'].sum()}")
print(f"Sum of number of Unknowns: {df._check['unknown_patents'].sum()}")
print(f"Number of patets with experiments extracted {len(df_check)}")
df_check.sort_values("Sum_error",ascending=False).head(20)

KeyError: 'nonprophetic_patents'

In [15]:
from utilities.nlp_processing import analyze_sentence_tense
def dic_to_dic_w_tense_test(doc_w_exp, threshold=0):
    dic = {}
    pattern = r"\(\d+\)\s*([A-Za-z0-9\-\(\)\{\},:;=\[\]\+\*\s\.\^\$\%]+(?:\.(?:sup|delta|Hz|NMR)[^\)]*)?)"

    for key, value in doc_w_exp.items():
        tense_counts = {"past": 0, "present": 0, "unknown": 0}

        if isinstance(value, list) and len(value) == 1:
            desc =  "".join(value[0]["content"]) # value[0]["title"] + "." +
            if len(desc) > threshold:
                tense = analyze_sentence_tense(desc)
                if tense != "unknown":
                    tense_counts[tense] += 1
                else:
                    matches = re.findall(pattern, desc)
                    if matches:
                        tense_counts["past"] += 1
                    else:
                        tense_counts["unknown"] += 1
                dic[key] = tense_counts

        elif isinstance(value, list) and len(value) > 1:
            for ls in value:
                desc = "".join(ls["content"]) #ls["title"] + "." + 
                if len(desc) > threshold:
                    if len(desc) > 0:
                        tense = analyze_sentence_tense(desc)

                        if tense != "unknown":
                            tense_counts[tense] += 1
                        else:
                            matches = re.findall(pattern, desc)
                            if matches:
                                tense_counts["past"] += 1
                            else:
                                tense_counts["unknown"] += 1
            dic[key] = tense_counts

        # elif isinstance(value, dict):
        #     print(value)
        #     for ex, desc in value.items():
        #         if len(desc) > threshold:
        #             tense = analyze_sentence_tense(desc)
        #             if tense != "unknown":
        #                 tense_counts[tense] += 1
        #             else:
        #                 matches = re.findall(pattern, desc)
        #                 if matches:
        #                     tense_counts["past"] += 1
        #                 else:
        #                     tense_counts["unknown"] += 1
        #     dic[key] = tense_counts
        # else:
        #     print(type(value))
        #     print(value)
        #     print(key)

    return dic

In [17]:
import pprint as pp
pp.pprint(merged[merged.patentnumber == "7723468"].xml.values[0])

('<?xml version="1.0" encoding="UTF-8"?>\n'
 '<!DOCTYPE sequence-cwu SYSTEM "us-sequence-listing-2004-03-09.dtd" [ ]>\n'
 '<sequence-cwu id="SEQLST-1">\n'
 '<publication-reference>\n'
 '<document-id>\n'
 '<country>US</country>\n'
 '<doc-number>07723468</doc-number>\n'
 '<kind>B2</kind>\n'
 '<date>20100525</date>\n'
 '</document-id>\n'
 '</publication-reference>\n'
 '<number>1</number>\n'
 '<sequence-list-new-rules>\n'
 '<s100>\n'
 '<s160>7</s160>\n'
 '</s100>\n'
 '<s200>\n'
 '<s210>1</s210>\n'
 '<s211>18</s211>\n'
 '<s212>PRT</s212>\n'
 '<s213>Artificial Sequence</s213>\n'
 '<s220>\n'
 '<s223>Synthesized Sequence</s223>\n'
 '</s220>\n'
 '</s200>\n'
 '<s400> 1\n'
 'Xaa Cys Arg Arg Leu Cys Tyr Lys Gln Arg Cys Val Thr Tyr Cys Arg\n'
 ' 1               5                  10                  15\n'
 'Gly Xaa\n'
 '</s400>\n'
 '<s200>\n'
 '<s210>2</s210>\n'
 '<s211>18</s211>\n'
 '<s212>PRT</s212>\n'
 '<s213>Acanthoscurria gomesiana</s213>\n'
 '<s220>\n'
 '<s221>\n'
 'VARIANT\n'
 '</s221>\n'
 '

In [143]:
from bs4 import BeautifulSoup

def extract_examples_start_w_word(xml_siblings):
    examples = []
    current_example = None
    in_example = False

    for tag in xml_siblings:
        if tag.name == "heading":
            if (
                tag.text.strip().lower().startswith("example")
                or tag.text.strip().lower().startswith("experiment")
                or tag.text.strip().lower().startswith("test")
                or tag.text.strip().lower().startswith("trial")
                or "test" in tag.text.strip().lower()
                or "experiment" in tag.text.strip().lower()
                or "example" in tag.text.strip().lower()
                or "trial" in tag.text.strip().lower()
            ):
                in_example = True
                current_example = {
                    "number": tag.text.strip(),
                    "title": xml_siblings[xml_siblings.index(tag) + 1].text.strip(),
                    "content": [],
                }
                examples.append(current_example)
        elif tag.name == "heading" and (
            tag.text.strip().lower().startswith("example")
            or tag.text.strip().lower().startswith("experiment")
            or tag.text.strip().lower().startswith("test")
            or tag.text.strip().lower().startswith("trial")
            or "test" in tag.text.strip().lower()
            or "experiment" in tag.text.strip().lower()
            or "example" in tag.text.strip().lower()
            or "trial" in tag.text.strip().lower()
        ):
            in_example = False
        # else:
        #     # If we hit any other heading, stop collecting content
        #     in_example = False
        elif in_example and current_example is not None:
            current_example["content"].append(tag.text.strip())

    return examples
num_of_paterns = 5000
mae = 0
highest_difference = 0
found_heading = 0
not_found_heading = 0
gib = 0
short = 0
mostdifss = []
doc_w_exp = {}

for row in df[:10000].iterrows():
    xml = row[1]["xml"]
    if len(xml)>2000:
        s_tags = re.findall(r'<s\d+>.*?</s\d+>', xml)
        if len(s_tags) > 0:
            #print(f"Patent {row[1]['patentnumber']} is gibberish")
            gib+=1
        else:
            heading = extract_experiments_w_heading(xml)
            #janetsnumexamples = row[1]["prophetic"] + row[1]["nonprophetic"]

            if heading:
                found_heading += 1
                examples = extract_examples_start_w_word(heading[0].find_next_siblings())
                if len(examples)==0:
                    soup = BeautifulSoup(xml, 'xml')
                    siblings = soup.find_all(['heading', 'p'])
                    examples = extract_examples_start_w_word(siblings)
                numexamples = len(examples)
            else:
                not_found_heading += 1
                soup = BeautifulSoup(xml, 'xml')
                siblings = soup.find_all(['heading', 'p'])
                examples = extract_examples_start_w_word(siblings)
                numexamples = len(examples)



            # #difference = abs(numexamples-janetsnumexamples)
            # mae += difference
            # if difference > 0:
            #     mostdifss.append([difference, row[1]["patentnumber"]])
            # if difference > highest_difference:
            #     highest_difference = difference
            #     highest_difference_patent = row[1]["patentnumber"]
            if len(examples)>0:
                doc_w_exp[row[1]["patentnumber"]] = examples

    else:
        short+=1
        #print(f"skipping {row[1]['patentnumber']}, patent is too short")
        
        
print(f"found_heading: {found_heading}, not_found_heading: {not_found_heading}, gibberish: {gib},too_short: {short}")
print(f"{num_of_paterns} patents MAE: {mae/num_of_paterns}, total_error: {mae}, highest_difference: {highest_difference}, highest_difference_patent: ") #highest_difference_patent

found_heading: 5386, not_found_heading: 2661, gibberish: 1731,too_short: 222
5000 patents MAE: 0.0, total_error: 0, highest_difference: 0, highest_difference_patent: 


In [209]:
a = []
num_examples = 0
for k, pat_exs in doc_w_exp.items():
    num_examples += len(pat_exs)
    for ex in pat_exs:
        ola = "".join(ex["content"]) + ex["title"]
        if len(ola) == 0:
            #print(ex)
            continue
        a.append({
            'patent_number': k,
            'content_number': ex["number"]
        })
df_ex_types = pd.DataFrame(a)
val_counts= df_ex_types.content_number.value_counts()
valc_df = pd.DataFrame(val_counts).reset_index()

# Group by type and sum counts
def get_example_type(content_number):
    content_lower = content_number.lower()
    for prefix in start_w:
        if content_lower.startswith(prefix):
            return prefix
    for special in ls:
        if special in content_lower:
            return special
    return "other"

# Extract first word from each content_number
def get_first_word(text):
    return text.lower().split()[0]

# Get common prefixes (appearing more than 100 times)
common_prefixes = (df_ex_types['content_number']
                  .apply(get_first_word)
                  .value_counts()
                  .loc[lambda x: x > 100])

start_w = common_prefixes.index.tolist()
print("Discovered prefixes:")
print(common_prefixes/df_ex_types.shape[0] * 100)

# Function to detect patterns that occur in middle of text
def find_common_phrases(df, min_count=50):
    # Get all content numbers as lowercase
    texts = df['content_number'].str.lower()
    # Find phrases with 2 or more words that appear frequently
    phrases = texts[texts.str.contains(' .+ ')]  # Contains at least 2 spaces
    common_phrases = phrases.value_counts().loc[lambda x: x > min_count]
    return common_phrases.index.tolist()

ls = find_common_phrases(df_ex_types)
print("\nDiscovered special phrases:")
print(ls)

# Rest of your grouping code remains the same
valc_df['type'] = valc_df['content_number'].apply(get_example_type)
type_counts = valc_df.groupby('type')['count'].sum()

Discovered prefixes:
content_number
example         82.320069
comparative      4.013693
examples         2.527713
reference        2.359714
synthesis        1.443805
preparation      1.299003
production       0.533519
experimental     0.504699
industrial       0.484314
test             0.409101
preparative      0.228450
synthetic        0.205957
formulation      0.179948
intermediate     0.103330
biological       0.098409
application      0.090677
experiment       0.075916
Name: count, dtype: float64

Discovered special phrases:
['comparative example 1', 'comparative example 2', 'comparative example 3', 'comparative example 4', 'comparative example 5', 'comparative example 6', 'reference example 1', 'comparative example 7', 'synthesis example 1', 'synthesis example 2', 'reference example 2', 'comparative example 8', 'synthesis example 3', 'test example 1', 'preparation example 1', 'comparative example 9', 'reference example 3', 'synthesis example 4', 'experimental example 1', 'test exa

In [180]:
# For 10000 patents in 2015:

# ---------------
# Starts with
# comparative example :  1035
# example example :  6682
# reference example :  158
# synthesis example :  156
# test example :  103
# ---------------
# Ohter
# industrial applicability :  663

In [182]:
print("\nUnique content numbers:", len(df_ex_types['content_number'].unique()))
print("\nContent number frequencies:")
print(df_ex_types['content_number'].value_counts())

print("\nUnique patent numbers:", len(df_ex_types['patent_number'].unique()))
print("\nPatent number frequencies:")
print(df_ex_types['patent_number'].value_counts())


Unique content numbers: 23329

Content number frequencies:
content_number
Example 1                       6682
Example 2                       6024
Example 3                       5260
Example 4                       4561
Example 5                       3855
                                ... 
Example #N.1.1                     1
Example #O.1.1                     1
Example #X.1.1                     1
Example #38                        1
VI. Industrial Applicability       1
Name: count, Length: 23329, dtype: int64

Unique patent numbers: 7965

Patent number frequencies:
patent_number
8952157    1446
8987441    1040
8969587     996
8987242     901
9006265     800
           ... 
9034169       1
8956508       1
8956528       1
8956536       1
8986522       1
Name: count, Length: 7965, dtype: int64


INDUSTRIAL APPLICABILITY  :  30


In [13]:
print(f"Number of patents with difference > 0: {len(mostdifss)}")
mostdifss

Number of patents with difference > 0: 104


[[4.0, '8926548'],
 [1.0, '8926595'],
 [1.0, '8926732'],
 [1.0, '8926824'],
 [6.0, '8926862'],
 [4.0, '8926864'],
 [1.0, '8926872'],
 [1.0, '8926955'],
 [11.0, '8926966'],
 [7.0, '8926979'],
 [1.0, '8927015'],
 [1.0, '8927023'],
 [5.0, '8927036'],
 [2.0, '8927098'],
 [1.0, '8927150'],
 [2.0, '8927157'],
 [1.0, '8927236'],
 [4.0, '8927254'],
 [1.0, '8927429'],
 [1.0, '8927469'],
 [1.0, '8927479'],
 [1.0, '8927536'],
 [1.0, '8927541'],
 [1.0, '8927546'],
 [2.0, '8927568'],
 [4.0, '8927583'],
 [6.0, '8927588'],
 [3.0, '8927596'],
 [1.0, '8927617'],
 [1.0, '8927619'],
 [2.0, '8927631'],
 [1.0, '8927642'],
 [1.0, '8927647'],
 [1.0, '8927678'],
 [1.0, '8927681'],
 [2.0, '8927683'],
 [1.0, '8927698'],
 [7.0, '8927710'],
 [15.0, '8927721'],
 [1.0, '8927725'],
 [1.0, '8927730'],
 [1.0, '8927738'],
 [1.0, '8927746'],
 [1.0, '8927750'],
 [1.0, '8927777'],
 [1.0, '8927781'],
 [1.0, '8927801'],
 [9.0, '8927811'],
 [1.0, '8932406'],
 [3.0, '8932470'],
 [4.0, '8932480'],
 [2.0, '8932491'],
 [2.0, '89

In [15]:
# highest difference
df2[df2.patentnumber == "8933072"][["patentnumber","prophetic","nonprophetic"]]

patentnumber  prophetic  nonprophetic
8285      8933072        0.0           9.0

In [ ]:
import numpy as np
final = dic_to_dic_w_tense_test(doc_w_exp,threshold=0)
df_final = pd.DataFrame(final).T.reset_index()
df_final.columns = ["patentnumber","past","present","Unknown"]
df_final["patentnumber"] = df_final["patentnumber"].apply(remove_leadiong_zeros)
df_check = df_final.merge(merged,on= "patentnumber",how="left")[["patentnumber","present","past","Unknown","prophetic","nonprophetic","allprophetic","someprophetic"]]
#df_check["past"] = df_check["past"] + df_check["Unknown"]
df_check["Total_Extracted"] = df_check["past"] + df_check["present"]  #+ df_check["Unknown"]
df_check["Total_Freilich"] = df_check["prophetic"] + df_check["nonprophetic"]
df_check["prophetic_error"] = np.sqrt((df_check["prophetic"] - df_check["present"])**2)
df_check["nonprophetic_error"] = np.sqrt((df_check["nonprophetic"] - df_check["past"])**2)
df_check["Total_Mean_error"] = np.sqrt((df_check["Total_Freilich"] - df_check["Total_Extracted"])**2)
df_check["Sum_error"] = df_check["prophetic_error"] + df_check["nonprophetic_error"] + df_check["Total_Mean_error"]
print(f"Number of exact matches: {len(df_check[(df_check.prophetic_error ==0) & (df_check.nonprophetic_error ==0) & (df_check.Total_Mean_error ==0)])} out of {len(df_check)} , Percentage: {len(df_check[(df_check.prophetic_error ==0) & (df_check.nonprophetic_error ==0) & (df_check.Total_Mean_error ==0)])/len(df_check)*100}")
print(f"Number of exact num of patent extracted: {len(df_check[(df_check.Total_Extracted == df_check.Total_Freilich)])} out of {len(df_check)}")
print(f"Avg Total error: {df_check['Total_Mean_error'].mean()}, num of corrects:  {df_check[df_check['Total_Mean_error'] == 0].shape[0]}")
print(f"Avg Total prophetic error: {df_check['prophetic_error'].mean()}, num of corrects: {df_check[df_check['prophetic_error'] == 0].shape[0]}")
print(f"Avg Total nonprophetic error: {df_check['nonprophetic_error'].mean()}, num of corrects: {df_check[df_check['nonprophetic_error'] == 0].shape[0]}")
print(f"{df_check['Sum_error'].sum()}")
print(f"Sum of number of Unknowns: {df_check['Unknown'].sum()}")
print(f"Number of patets with experiments extracted {len(df_check)}")
df_check.sort_values("Sum_error",ascending=False).head(20)


# 1000
# Number of exact matches: 577 out of 799 , Percentage: 72.21526908635795
# Number of exact num of patent extracted: 694 out of 799
# Avg Total error: 0.5181476846057572, num of corrects:  694
# Avg Total prophetic error: 0.5857321652065082, num of corrects: 652
# Avg Total nonprophetic error: 0.9887359198998749, num of corrects: 587
# 1672.0
# Sum of number of Unknowns: 0
# Number of patets with experiments extracted 799

# 5000
# Number of exact matches: 2870 out of 3971 , Percentage: 72.27398640141023
# Number of exact num of patent extracted: 3461 out of 3971
# Avg Total error: 0.5560312263913372, num of corrects:  3461
# Avg Total prophetic error: 0.8645177537144296, num of corrects: 3209
# Avg Total nonprophetic error: 1.2976580206497104, num of corrects: 2942
# 10794.0
# Sum of number of Unknowns: 0
# Number of patets with experiments extracted 3971

SyntaxError: invalid syntax (2897131579.py, line 34)

In [ ]:
# patent 9102614 injanets says 4 examples but there are 161
# patent 9102599 in janets says 1 ,theres more than 1 actually 
# Some xml files are made with error : e.g. these dont have much information : skipping 9102601, 9102662,9102692
# these have gibberish: 9102724,9102705,9102727,9102628

In [ ]:
def process_siblings_test(xml_siblings):
    examples = []

    # Find all matching headings directly from xml_siblings
    example_headings = [
        tag
        for tag in xml_siblings
        if tag.name == "heading"
        and any(
            keyword in tag.text.strip().lower().replace(" ", "")
            for keyword in ["example", "experiment", "test"]
        )
        # and not any(
        #     excluded in tag.text.strip().lower().replace(" ", "")
        #     for excluded in ["reference", "preparation"]
        # )
    ]

    for heading in example_headings:
        current_content = []
        idx = xml_siblings.index(heading)

        # Get title from next heading if available
        title = ""
        if idx + 1 < len(xml_siblings) and xml_siblings[idx + 1].name == "heading":
            title = xml_siblings[idx + 1].text.strip()

        # Collect content until next example heading
        i = idx + 1
        while i < len(xml_siblings):
            if (
                xml_siblings[i].name == "heading"
                and any(
                    keyword in xml_siblings[i].text.strip().lower().replace(" ", "")
                    for keyword in ["example", "experiment", "test"]
                )
                # and not any(
                #     excluded in xml_siblings[i].text.strip().lower().replace(" ", "")
                #     for excluded in ["reference", "preparation"]
                # )
            ):
                break
            if xml_siblings[i].name == "p":
                current_content.append(xml_siblings[i].text.strip())
            i += 1

        examples.append(
            {"number": heading.text.strip(), "title": title, "content": current_content}
        )

    return examples if examples else None





In [6]:
def extract_examples_from_heading(heading):
    extracted_examples = []
    example_start_w = process_siblings(heading.find_next_siblings())
    if example_start_w:
        if not example_start_w[0]["content"]:
            extracted_ex_w_word = extract_examples_start_w_word(heading.find_next_siblings())
            if extracted_ex_w_word:
                if isinstance(extracted_ex_w_word, list):
                    if extracted_ex_w_word and len(extracted_ex_w_word[0]["content"])>0:
                        extracted_examples.append(extracted_ex_w_word)
                elif extracted_ex_w_word["content"]:
                    extracted_examples.append(extracted_ex_w_word)
        else:
            if len(example_start_w[0]["content"])>0:
                extracted_examples.append(example_start_w)
    else:
        extracted_ex_w_word = extract_examples_start_w_word(heading.find_next_siblings())
        if extracted_ex_w_word:
            if isinstance(extracted_ex_w_word, list):
                if extracted_ex_w_word and len(extracted_ex_w_word[0]["content"])>0:
                    extracted_examples.append(extracted_ex_w_word)
            elif extracted_ex_w_word["content"]:
                if len(extracted_ex_w_word["content"])>0:
                    extracted_examples.append(extracted_ex_w_word)
        else:
            num_dot_examples = extract_num_dot_examples(str(heading.find_next_siblings()))
            if num_dot_examples:
                extracted_examples.append(num_dot_examples)
    return extracted_examples

def extract_examples(xml):
    heading = extract_experiments_w_heading(xml)
    if heading:
        if len(heading) > 1:
            print(f"has more than 1 Examples heading, {i}")
        elif len(heading) == 1:
            return extract_examples_from_heading(heading[0])
        else:
            extracted_ex_w_word = extract_examples_w_word(xml)
            if extracted_ex_w_word:
                if isinstance(extracted_ex_w_word, list):
                    if extracted_ex_w_word and extracted_ex_w_word[0]["content"]:
                        return [extracted_ex_w_word]
                elif extracted_ex_w_word["content"]:
                    return [extracted_ex_w_word]
            else:
                example_start_w = process_siblings(heading[0].find_next_siblings())
                if example_start_w and example_start_w[0]["content"]:
                    return [example_start_w]
    else: 
        example_start_w = extract_examples_w_word(xml)
        if example_start_w:
            return [example_start_w]
        else:
            num_dot_examples = extract_num_dot_examples(xml)
            if num_dot_examples:
                return [num_dot_examples]
    return []

doc_w_exp = {}
for i, xml in enumerate(merged["xml"][:1000], start=1):
    if i % 1000 == 0:
        print(f"{i}/{len(merged.xml)} so far found {len(doc_w_exp)} docs with experiments")
    
    extracted_examples = extract_examples(xml)
    if extracted_examples:
        doc_w_exp[find_doc_number(xml)[0]] = extracted_examples

has more than 1 Examples heading, 349
has more than 1 Examples heading, 388
has more than 1 Examples heading, 689
has more than 1 Examples heading, 768
has more than 1 Examples heading, 781
has more than 1 Examples heading, 814
1000/25081 so far found 771 docs with experiments


In [73]:
def extract_examples_from_heading(heading):
    length_threshold = 30
    extracted_examples = []
    example_start_w = process_siblings(heading.find_next_siblings())
    if example_start_w:
        if not example_start_w[0]["content"]:
            extracted_ex_w_word = extract_examples_start_w_word(heading.find_next_siblings())
            if extracted_ex_w_word:
                if isinstance(extracted_ex_w_word, list):
                    extracted_ex_w_word = [ex for ex in extracted_ex_w_word if len(ex["content"])>length_threshold]
                    extracted_examples.append(extracted_ex_w_word)
                elif extracted_ex_w_word["content"]>length_threshold:
                    extracted_examples.append(extracted_ex_w_word)
        else:
            if isinstance(example_start_w, list):
                example_start_w = [ex for ex in example_start_w if len(ex["content"])>length_threshold]
                extracted_examples.append(example_start_w)
    else:
        extracted_ex_w_word = extract_examples_start_w_word(heading.find_next_siblings())
        if extracted_ex_w_word:
            if isinstance(extracted_ex_w_word, list):
                extracted_ex_w_word = [ex for ex in extracted_ex_w_word if len(ex["content"])>length_threshold]
                extracted_examples.append(extracted_ex_w_word)
            elif len(extracted_ex_w_word["content"])>length_threshold:
                extracted_examples.append(extracted_ex_w_word)
        else:
            num_dot_examples = extract_num_dot_examples(str(heading.find_next_siblings()))
            if isinstance(num_dot_examples, list):
                num_dot_examples = [ex for ex in num_dot_examples if len(ex["content"])>length_threshold]
                extracted_examples.append(num_dot_examples)
            elif isinstance(num_dot_examples, dict):
                num_dot_examples = [ex for ex in num_dot_examples.items() if len(ex["content"])>length_threshold]

                extracted_examples.append(num_dot_examples)
    return extracted_examples

def extract_examples(xml):
    length_threshold = 30
    heading = extract_experiments_w_heading(xml)
    if heading:
        if len(heading) > 1:
            print(f"has more than 1 Examples heading, {i}")
        elif len(heading) == 1:
            return extract_examples_from_heading(heading[0])
        else:
            pass
        # else:
        #     extracted_ex_w_word = extract_examples_w_word(xml)
        #     if extracted_ex_w_word:
        #         if isinstance(extracted_ex_w_word, list):
        #             extracted_ex_w_word = [ex for ex in extracted_ex_w_word if len(ex["content"])>length_threshold]
        #             return extracted_ex_w_word if extracted_ex_w_word
        #         elif extracted_ex_w_word["content"]>length_threshold:
        #             return extracted_ex_w_word if extracted_ex_w_word
        #     else:
        #         example_start_w = process_siblings(heading[0].find_next_siblings())
        #         if isinstance(example_start_w, list):
        #             example_start_w = [ex for ex in example_start_w if len(ex["content"])>length_threshold]
        #             return example_start_w if example_start_w
    else: 
        example_start_w = extract_examples_w_word(xml)
        if isinstance(example_start_w, list):
            example_start_w = [ex for ex in example_start_w if len(ex["content"])>length_threshold]
            return example_start_w
        else:
            num_dot_examples = extract_num_dot_examples(xml)
            if isinstance(num_dot_examples, list):
                num_dot_examples = [ex for ex in num_dot_examples if len(ex["content"])>length_threshold]
                return num_dot_examples

doc_w_exp = {}
for i, xml in enumerate(merged["xml"][:1000], start=1):
    if i % 100 == 0:
        print(f"{i}/{len(merged.xml)} so far found {len(doc_w_exp)} docs with experiments")
    
    extracted_examples = extract_examples(xml)
    if extracted_examples:
        doc_w_exp[find_doc_number(xml)[0]] = extracted_examples



100/25081 so far found 54 docs with experiments
200/25081 so far found 111 docs with experiments
300/25081 so far found 169 docs with experiments
has more than 1 Examples heading, 349
has more than 1 Examples heading, 388
400/25081 so far found 225 docs with experiments
500/25081 so far found 294 docs with experiments
600/25081 so far found 352 docs with experiments
has more than 1 Examples heading, 689
700/25081 so far found 395 docs with experiments
has more than 1 Examples heading, 768
has more than 1 Examples heading, 781
800/25081 so far found 463 docs with experiments
has more than 1 Examples heading, 814
900/25081 so far found 522 docs with experiments
1000/25081 so far found 575 docs with experiments


In [93]:

doc_w_exp = {}
for i, xml in enumerate(merged[:1999].xml.values, start=1):
    if i % 1000 == 0:
        print(f"{i}/{len(merged.xml)} so far found {len(doc_w_exp)} docs with experiments")
    
    extracted_examples = extract_examples(xml)
    if extracted_examples:
        doc_w_exp[find_doc_number(xml)[0]] = extracted_examples

has more than 1 Examples heading, 349
has more than 1 Examples heading, 388
has more than 1 Examples heading, 689
has more than 1 Examples heading, 768
has more than 1 Examples heading, 781
has more than 1 Examples heading, 814
1000/25081 so far found 769 docs with experiments
has more than 1 Examples heading, 1054
has more than 1 Examples heading, 1417
has more than 1 Examples heading, 1475
has more than 1 Examples heading, 1810


In [5]:
# save_as_pickle(doc_w_exp,"doc_w_exp_new_algorithm.pkl")
doc_w_exp = load_from_pickle("doc_w_exp_new_algorithm.pkl")

# save_as_json(doc_w_exp, "doc_w_exp_2015.json")
# doc_w_exp = read_json("../data/doc_w_exp_2015.json")

Loaded 18977 patents from doc_w_exp_new_algorithm.pkl


In [10]:
import re
import html
import unicodedata
from bs4 import BeautifulSoup

def clean_text_updated(text):
    """
    Clean text by removing HTML tags, special characters, extra spaces,
    and normalizing the content while keeping meaningful punctuation.

    Args:
        text (str): Input text to clean

    Returns:
        str: Cleaned text
    """
    if not isinstance(text, str):
        return ""

    # Decode HTML entities & remove HTML tags
    text = html.unescape(text)
    text = BeautifulSoup(text, "html.parser").get_text()

    # Normalize Unicode characters (e.g., é → e)
    text = unicodedata.normalize("NFKC", text)

    # Remove unwanted special characters but keep punctuation
    text = re.sub(r"[^\w\s.,!?'\-]", "", text)

    # Normalize spaces: remove multiple spaces, newlines, and tabs
    text = re.sub(r"\s+", " ", text).strip()

    return text

import nltk
from nltk.tokenize import word_tokenize
from nltk.tag import pos_tag

def check_tense_nltk_updated(sentence):
    words = word_tokenize(sentence)
    tagged = pos_tag(words)

    # Define tense categories
    past = {"VBD", "VBN"}
    present = {"VB", "VBG", "VBP", "VBZ"}
    future = {"MD"}

    tenses = {"past": 0, "present": 0, "future": 0}

    for i, (word, tag) in enumerate(tagged):
        # Count past tense words
        if tag in past:
            tenses["past"] += 1

        # Count present tense words
        elif tag in present:
            tenses["present"] += 1

        # Future tense handling
        elif tag in future:
            if word.lower() in {"will", "shall"}:
                # Ensure 'will' or 'shall' is followed by a verb (to confirm future tense)
                if i + 1 < len(tagged) and tagged[i + 1][1] in {"VB", "VBP"}:
                    tenses["future"] += 1

    return max(tenses, key=tenses.get) if max(tenses.values()) > 0 else "Unknown"



In [ ]:
import re
from utilities.nlp_processing import check_tense_nltk,clean_text

def tense_with_pattern(description):
    future_pattern = r'\b(?:will|would|shall|should)\b'
    past_pattern = r'\b(?:was|were|had|did)\b'

    future_matches = re.findall(future_pattern, description)
    past_matches = re.findall(past_pattern, description)
    if future_matches and past_matches:
        if len(future_matches) > len(past_matches):
            return "present"
        else:
            return "past"
    if future_pattern or past_pattern:
        if future_pattern:
            return "present"
        else:
            return "past"
    return "Unkown"


def dic_to_dic_w_tense(doc_w_exp,threshold = 50):
    dic = {}
    pattern = r'\(\d+\)\s*([A-Za-z0-9\-\(\)\{\},:;=\[\]\+\*\s\.\^\$\%]+(?:\.(?:sup|delta|Hz|NMR)[^\)]*)?)'
    pattern2 = r'\(\d+\)\s*(?:[A-Za-z]+\s*\d*\-?[A-Za-z]*[\(\{][^)]*[\)\}][^;]*|(?:\.\d*H\-?NMR[^;]*|\d+\.[a-zA-Z]*\d*))'


    for key, value in doc_w_exp.items():
        tense_counts = {"past": 0, "present": 0,"Unknown":0}
        
        if isinstance(value, list) and len(value) == 1 and len(value[0][0]["content"]) > threshold:
            desc = value[0][0]["title"] + "." + "".join(value[0][0]["content"])
            tense = check_tense_nltk_updated(clean_text(desc))
            if tense !=  "Unknown":
                tense_counts[tense] += 1
            else:
                matches = re.findall(pattern, desc)
                if matches:
                    tense_counts["past"] += 1
                else:
                    tense_counts["Unknown"] += 1
            dic[key] = tense_counts


        elif isinstance(value[0], list) and len(value[0]) > 1:
            for ls in value[0]:
                if len(ls["content"]) > threshold:
                    desc = "".join(ls["content"])
                    tense = check_tense_nltk_updated(clean_text(desc))
                    if tense !=  "Unknown":
                        tense_counts[tense] += 1
                    else:
                        matches = re.findall(pattern, desc)
                        if matches:
                            tense_counts["past"] += 1
                        else:
                            tense_counts["Unknown"] += 1
            dic[key] = tense_counts

        elif isinstance(value[0], dict):
            for ex, desc in value[0].items():
                if len(desc) > threshold:
                    tense = check_tense_nltk_updated(clean_text(desc))
                    if tense !=  "Unknown":
                        tense_counts[tense] += 1
                    else:
                        matches = re.findall(pattern, desc)
                        if matches:
                            tense_counts["past"] += 1
                        else:
                            tense_counts["Unknown"] += 1
            dic[key] = tense_counts
        else:
            print(type(value[0]))
            print(value[0])

    return dic

import re

def dic_to_dic_w_unknown_tense(doc_w_exp):
    dic = {}
    pattern = r'\(\d+\)\s*([A-Za-z0-9\-\(\)\{\},:;=\[\]\+\*\s\.\^\$\%]+(?:\.(?:sup|delta|Hz|NMR)[^\)]*)?)'
    pattern2 = r'\(\d+\)\s*(?:[A-Za-z]+\s*\d*\-?[A-Za-z]*[\(\{][^)]*[\)\}][^;]*|(?:\.\d*H\-?NMR[^;]*|\d+\.[a-zA-Z]*\d*))'

    for key, value in doc_w_exp.items():
        unknown_experiments = []  # Initialize a list to store the experiments classified as Unknown
        
        if isinstance(value[0], list) and len(value[0]) == 1:
            desc = value[0][0]["title"] + "." + "".join(value[0][0]["content"])
            tense = check_tense_nltk_updated(clean_text(desc))
            if tense == "Unknown":
                unknown_experiments.append(desc)

        elif isinstance(value[0], list) and len(value[0]) > 1:
            for ls in value[0]:
                desc = ls["title"] + "." + "".join(ls["content"])
                tense = check_tense_nltk_updated(clean_text(desc))
                if tense == "Unknown":
                    unknown_experiments.append(desc)

        elif isinstance(value[0], dict):
            for ex, description in value[0].items():
                tense = check_tense_nltk_updated(clean_text(description))
                if tense == "Unknown":
                    unknown_experiments.append(description)

        # Only add to the dictionary if there are "Unknown" experiments
        if unknown_experiments:
            dic[key] = unknown_experiments

    return dic


In [105]:
# 8946300 example 1-5
# 08946296 example 2 onwards 
# 08932484 examples 2 to 11


# 08933074:  
# Synthesis of 5,5-dimethyl-4-(pyridin-4-yl)-3-(4-(quinolin-2-ylmethoxy)phenyl) furan-2(5H)-one (Example 23)
# 5,5-dimethyl-4-(pyridin-4-yl)-3-(4-(quinolin-2-ylmethoxy)phenyl) furan-2(5H)-one (Example 23)

# 08945522:
# Example 2
# (69) 3-(3-(Triethoxysilyl)propylamino)phenyl benzoate and
# (140) Example 9
# (141) 4-(3-(Triethoxysilyl)propylamino)phenyl benzoate and
# (142) Example 10
# (143) 4-(Bis(3-(triethoxysilyl)propyl)amino)phenyl benzoate



SyntaxError: invalid syntax (1140748136.py, line 1)

In [ ]:
[8969641,8926966,8951515,8940436,9023999,9011956,9018396,9213027,9108964,9084992]#examples our algorithm didnt find but freilich did 
# many dont have an overarching examples section but have example 1, example 2 etc
# check these xml, the text has examples section:9194008,9018332,9054322,9150561,9213027, 

In [10]:
# import pprint as p
# p.pprint(merged[merged.patentnumber == "9018332"]["xml"].values[0])

# ERRORS IN FREILICH DATASET:
patentnumber 8957069: in freilich 108.0	nonproph and 0 prophetic but there are many more examples than 108, our algorithm: 267 nonprophetic and 1 prophetic and 3 unknown
patent number 8946443:we extracted 271	freilich: 108, this patent has refrtence examples, are we extracting these or not? 
patent number 8952010: same issue do we take refrence examples? 
patent 8933099: there are far more than 33 examples in this patent but freilich says 33, our algorithm found 131
what to do with Preparation examples?? e.g 8962612 in freilich has 6 we got 70.

# improvement for our algorithm
1. remove duplicates somehow, e.g in patent number 08987242, there some examples are extracted twice just because the name e.g example aa137 appears multiple times.
2. better classification of the patents. e.g patent 8987295 has 442, we extract the same number as frelich but we clasify differently.
3. 